### ISSUE TITLE
the mAP is different between validation in training and directly validate

### ISSUE THREAD

**Issue by NOTGOOOOD** [ultralytics/yolov5#13629]
Hello, 

when I trained yolov5 (using `yolov5n.pt` as pre-training weights) on the COCO dataset,  I found a strange problem that the mAP is very low after a few training epochs. Even I freeze all trainable weights, the situation remains consistent. But, If I run `val.py` to validate directly, then the accuracy is correct.

**Detail:**

First, I run the `train.py` 
```Python
python train.py --weights yolov5n.pt --data data/coco.yaml
```
and validate the model **before training**, 
```Python
  callbacks.run("on_train_start")
  ema.update_attr(model, include=["yaml", "nc", "hyp", "names", "stride", "class_weights"])
  results, maps, _ = validate.run(
      data_dict,
      batch_size=batch_size // WORLD_SIZE * 2,
      imgsz=imgsz,
      half=amp,
      model=ema.ema,
      single_cls=single_cls,
      dataloader=val_loader,
      save_dir=save_dir,
      plots=False,
      callbacks=callbacks,
      compute_loss=compute_loss,
  )

  for epoch in range(start_epoch, epochs):
    model.train()
...
```
the results as follows
<img width="1315" height="376" alt="Image" src="https://github.com/user-attachments/assets/2baae82d-9b99-41c4-86a6-d83a18196946" />

And, then I run the `val.py`, 
```Python
python val.py --weights yolov5n.pt --data data/coco.yaml
```
you can see the correct result
<img width="1357" height="171" alt="Image" src="https://github.com/user-attachments/assets/5f6649fa-3c1f-4584-8eaa-3c395f14c579" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @NOTGOOOOD, thank you for bringing this to our attention and for your detailed description and screenshots! 🚀 This is an automated response to help get you started, and an Ultralytics engineer will follow up with you soon.

If this is a 🐛 Bug Report, please provide a **minimum reproducible example (MRE)** to help us debug the issue. This should include the specific commands you ran, any changes you made to the default configuration, and sample data or scripts if possible.

If this is a training ❓ Question, please share as much detail as possible, including dataset information, training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To set up your environment:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```
## Environments
YOLOv5 runs in the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py), and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience! One of our engineers will review and assist you soon. 😊

**Comment by pderrenger** [comment]
This discrepancy typically occurs due to differences in validation settings between training and standalone validation. The most common causes are different batch sizes (training uses `batch_size // WORLD_SIZE * 2` while val.py uses default batch size of 32) or model state differences (EMA vs non-EMA model).

Try running val.py with the same batch size used during training validation to isolate the issue: `python val.py --weights yolov5n.pt --data data/coco.yaml --batch-size X` where X matches your training validation batch size. Also ensure you're using the latest YOLOv5 version as this type of validation inconsistency has been addressed in recent updates.

**Comment by NOTGOOOOD** [comment]
Thanks for your reply! @pderrenger 

I've found the key reason, the activate layer have changed to `ReLU` by myself. 

This issue can be closed.

**Comment by pderrenger** [comment]
Great detective work @NOTGOOOOD! Activation function changes can definitely cause validation discrepancies since they alter the model's forward pass behavior and output distributions. Glad you tracked down the root cause - this type of architectural modification often has subtle effects that only show up during validation comparisons.
